File: phone/app/src/main/java/net/novucs/slither/Vector2i.java
package net.novucs.slither;

import android.graphics.Rect;

import java.util.Collection;

public class Vector2i {

    private final int x;
    private final int y;

    public Vector2i() {
        this(0, 0);
    }

    public Vector2i(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public Vector2i setX(int x) {
        return new Vector2i(x, y);
    }

    public int getY() {
        return y;
    }

    public Vector2i setY(int y) {
        return new Vector2i(x, y);
    }

    public Vector2i add(int x, int y) {
        return new Vector2i(this.x + x, this.y + y);
    }

    public Vector2i add(Vector2i other) {
        return add(other.x, other.y);
    }

    public Rect toBlock(int width, int height) {
        return new Rect(x * width, y * height, (x + 1) * width, (y + 1) * height);
    }

    public boolean isDefault() {
        return x == 0 && y == 0;
    }

    public boolean collides(Vector2i other) {
        return x == other.x && other.y == this.y;
    }

    public boolean collides(Collection<Vector2i> locations) {
        for (Vector2i other : locations) {
            if (collides(other)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "Vector2i{" +
                "x=" + x +
                ", y=" + y +
                '}';
    }
}

File: android/app/src/main/java/net/novucs/slither/Vector2i.kt
package net.novucs.slither

import android.graphics.Rect

/**
 * Represents a two dimensional vector of integers.
 */
data class Vector2i constructor(var x: Int = 0, var y: Int = 0) {

    val isDefault: Boolean
        get() = x == 0 && y == 0

    /**
     * Adds two vectors together.
     *
     * @param other the other vector to add to this.
     */
    fun add(other: Vector2i): Vector2i {
        return Vector2i(x + other.x, y + other.y)
    }

    /**
     * Translates the position to a one sized block to be rendered to the game
     * view.
     *
     * @param width the block width.
     * @param height the block height.
     */
    fun toBlock(width: Int, height: Int): Rect {
        return Rect(x * width, y * height, (x + 1) * width, (y + 1) * height)
    }

    /**
     * Checks if another vector is the same or collides with this.
     *
     * @param other the other vector.
     */
    fun collides(other: Vector2i): Boolean {
        return x == other.x && other.y == this.y
    }

    /**
     * Checks if any of a collection of locations collides with this.
     *
     * @param locations the locations to check against this.
     */
    fun collides(locations: Collection<Vector2i>): Boolean {
        return locations.any { collides(it) }
    }
}
